src/persist-client/src/impl/encoding.rs

```diff
@@ -87,140 +92,169 @@ where
     D: Codec64,
 {
     fn codec_name() -> String {
-        "StateRollupMeta".into()
+        "md5+protobuf[State]".into()
     }
 
     fn encode<B>(&self, buf: &mut B)
     where
         B: bytes::BufMut,
     {
-        let bytes = bincode::serialize(&StateRollupMeta::from(self))
-            .expect("unable to serialize BlobState");
-        buf.put_slice(&bytes);
+        // TODO: Move checksum to be a field on the proto instead. We can encode
+        // the proto, checksum'ing as we go, and then manually append it onto
+        // the end.
+        //
+        // TODO: Regardless of the above TODO, compute the checksum as we go and
+        // avoid this temp Vec.
+        let temp = self.into_proto().encode_to_vec();
+        buf.put_slice(&temp);
+        let checksum = ProtoStateRollup::md5_checksum(&temp);
+        buf.put_slice(&checksum);
     }
 
     fn decode<'a>(buf: &'a [u8]) -> Result<Self, String> {
-        let state: StateRollupMeta =
-            bincode::deserialize(buf).map_err(|err| format!("unable to decode state: {}", err))?;
-        // This map_err goes away when we do incremental state.
-        State::try_from(&state).map_err(|err| err.to_string())
+        let proto = ProtoStateRollup::decode_checksummed(buf)?;
+        // This match goes away when we do incremental state.
+        match State::try_from(proto) {
+            Ok(Ok(x)) => Ok(x),
+            Ok(Err(err)) => Err(err.to_string()),
+            Err(err) => Err(err.to_string()),
+        }
+    }
+}
+
+impl ProtoStateRollup {
+    // NB: This len is intentionally hardcoded (not derived from the md5 crate)
+    // so that a change to the crate can't break us. The compiler statically
+    // checks that they match because this const is used in the return type.
+    // (MD5 is not going to change, so this is all a moot point, but still
+    // better to be defensive.)
+    const CHECKSUM_LEN: usize = 16;
+
+    fn decode_checksummed(buf: &[u8]) -> Result<ProtoStateRollup, String> {
+        if buf.len() < Self::CHECKSUM_LEN {
+            return Err("missing/incomplete checksum".into());
+        }
+        let checksum_start = buf.len() - Self::CHECKSUM_LEN;
+        let (buf, checksum) = (&buf[..checksum_start], &buf[checksum_start..]);
+        if checksum != &Self::md5_checksum(&buf) {
+            return Err("checksum mismatch".into());
+        }
+
+        ProtoStateRollup::decode(buf).map_err(|err| err.to_string())
+    }
+
+    fn md5_checksum(buf: &[u8]) -> [u8; Self::CHECKSUM_LEN] {
```

Comment:
What's the reason behind us using md5? I know we've also used it before but I've recently read this article which argues against using these classic but slow algorithms: https://jolynch.github.io/posts/use_fast_data_algorithms/

Reply:
I don't have one, except "use xxhash" but just because that's what the article recommends.